crates/noirc_printable_type/src/lib.rs

```diff
@@ -0,0 +1,321 @@
+use std::{collections::BTreeMap, str};
+
+use acvm::{brillig_vm::brillig::Value, FieldElement};
+use iter_extended::vecmap;
+use regex::{Captures, Regex};
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(tag = "kind", rename_all = "lowercase")]
+pub enum PrintableType {
+    Field,
+    Array {
+        length: u64,
+        #[serde(rename = "type")]
+        typ: Box<PrintableType>,
+    },
+    SignedInteger {
+        width: u32,
+    },
+    UnsignedInteger {
+        width: u32,
+    },
+    Boolean,
+    Struct {
+        name: String,
+        fields: Vec<(String, PrintableType)>,
+    },
+    String {
+        length: u64,
+    },
+}
+
+impl PrintableType {
+    /// Returns the number of field elements required to represent the type once encoded.
+    fn field_count(&self) -> u32 {
+        match self {
+            Self::Field
+            | Self::SignedInteger { .. }
+            | Self::UnsignedInteger { .. }
+            | Self::Boolean => 1,
+            Self::Array { length, typ } => typ.field_count() * (*length as u32),
+            Self::Struct { fields, .. } => {
+                fields.iter().fold(0, |acc, (_, field_type)| acc + field_type.field_count())
+            }
+            Self::String { length } => *length as u32,
+        }
+    }
+}
+
+/// This is what all formats eventually transform into
+/// For example, a toml file will parse into TomlTypes
+/// and those TomlTypes will be mapped to Value
+#[derive(Debug, Clone, Serialize, PartialEq)]
+pub enum PrintableValue {
+    Field(FieldElement),
+    String(String),
+    Vec(Vec<PrintableValue>),
+    Struct(BTreeMap<String, PrintableValue>),
+}
+
+/// In order to display a `PrintableValue` we need a `PrintableType` to accurately
+/// convert the value into a human-readable format.
+pub enum PrintableValueDisplay {
+    Plain(PrintableValue, PrintableType),
+    FmtString(String, Vec<(PrintableValue, PrintableType)>),
+}
+
+#[derive(Debug, Error)]
+pub enum ForeignCallError {
```

Comment:
I see why you moved these here. If we ever decide to have compiler builtin foreign calls not related to printing we can always bring back the error enum into nargo